---
title: "R markdown of BeHo initial analyses"
output: pdf_document
date: "2023-10-16"
---

## Project description

Much evidence suggests that the gut microbiome influences animal behaviour. Nevertheless, many behavioural traits have not been investigated yet, in particular those unassociated with diseases. Thus, in this study, we infer the behaviour dominance in J:DO laboratory mice (mus musculus) to test if the gut microbiome plays a role herein during different experimental treatments. Over the course of 14 weeks, we measured dominance of 80 mice in total, 40 of each sex, using a tube test to rank the dominance within a cage. Five males or five females were housed together per cage, and a hierarchy of dominance was therefore created ranking the individuals from 1 to 5, where 1 signified “most dominant” and 5 as “most submissive”. We further wanted to explore if the gut microbiome affects dominance under different biological conditions, such as changes in temperature or diet. Hence, the mice were exposed to five specific treatments, all known to interfere with the gut microbiome, each following a four day resting period for the mice to restore the potential unbalance caused (e.g. physiological or gut microbial). The order of treatments was i) an exposure to heat (temperature set to 34 degrees celsius), ii) an exposure to cold (temperature set to 14 degrees celsius), iii) a dietary change, where tryptophan was removed, iv) an antibiotic and antifungal treatment, and lastly, v) a fecal microbiota transplantation, where, among other, the feces from the most dominant animal was given to the most submissive, and vice versa, within each cage. However, one cage per sex were not exposed to any changes, and thus, they serve as control cages to examine what the gut microbes may have had of an effect on dominance without any disturbances. 

Read more about the BeHo mice project:
https://docs.google.com/document/d/1Ra6K9tJBcnQmh23X5Pv3lZoZJmxm84B3S1uOH_diuWY/edit?usp=sharing

## Libraries are loaded, the environment set and files read in

The libraries loaded. Not all of them are used though.
library(tidyverse)
library(reshape2)
library(tidyr)
library(effects)
library(ggplot2)
library(sjPlot)
library(vioplot)
library(ggpubr)
library(PerformanceAnalytics)

Mice who died during the experiment before it ended have zeros (0) written as tube test results, which are misguiding. Thus, the zeroes are replaced with NA.

To explain the columns of the dataframes that might not be intuitive:

R1, R2, R3, R4 and R5 = the replicates of tube test per treatment. Each replicate represents a day. Hence, if a treatment has three replicates, then they were performed over three days. All cages were tested in one day and the repeated the next. The mice were tested in the last days of the experiment (except for FMT). For instance, in the diet treatment the last three days = three tube test replicates. One per day. Thus, the diet treatment lasted 8 days and the tube tests were conducted on day 6, 7 and 8. However, the FMT treatment there are five replicates in total, because we wanted to test if the dominance changes during the FMT treatment. Thus, we did tube tests FMT day 3, 6, 9, 10 and 11. Thus, FMT day 9-11 are the three consecutive replicates like the previous treatments.

Wins_per_rep = the number of wins an individual mouse has within a replicate. The number of wins can vary from 0 to 4. As all five mice within a cage "fight" each other. However, the number of wins don't necessarily dsitribute as 0, 1, 2, 3 and 4 for each mouse within a cage.

OPT = optimal conditions. The mice were household in 24 degrees celsius and no disturbances were made. Alias for this "treatment": Acclimation. Duration: 4 weeks.
HEAT = The experimental mice were exposed to 34 degrees celsius. Duration: 12 days.
COLD = The experimental mice were exposed to 14 degrees celsius. Duration: 12 days.
DIET = The experimental mice were exposed to a tryptophan depleted diet. Duration: 8 days.
ANTI = The experimental mice were exposed to antibiotics and anti-fungals to remove the existing gut microbes. The mice were given antibiotics for the first five days. Duration: 8 days.
FMT = The experimental mice were exposed to fecal microbiota transplantations (FMT). The FMTs were given the first two days. Duration: 11 days. 


```{r setup, include=FALSE}
#loading libraries, setting environment/session and reading in files
library(tidyverse)
library(reshape2)
library(tidyr)
library(effects)
library(ggplot2)
library(sjPlot)
library(vioplot)
library(ggpubr)
library(PerformanceAnalytics)
setwd("~/Documents/PhD 2020/BeHo:ZiBA study/Data analysis/Tube test data analysis")
tube_test_wins = read.csv("R_input_files/Mice-full-overview.csv", header = TRUE)

# placing NA for dead mice (C8 mouse 3065, C16 mouse 3183, C7 mouse 2730, C3 mouse 3083, C1 mouse 4176)
tube_test_wins[76, c(32:45,48:49)] = NA
tube_test_wins[38, c(32:45,48:49)] = NA
tube_test_wins[32, c(32:45,48:49)] = NA
tube_test_wins[14, c(32:45,48:49)] = NA
tube_test_wins[4,c(28:45,47:49)] = NA
#Removing irrelevant columns
tube_test_80 <- tube_test_wins [,-c(5,8,10,15:17,21:23, 27, 31, 35:37, 39, 42:50)]
#tube_test_80 <- tube_test_wins [,-c(5,8,10,39,42:44)] #I would like it changed into this over time, but for now I can't. It will ruin the "Replicates" and "levels" commands.

#Renaming columns without the many dots
names(tube_test_80)[2] <- "Arrive_weight"
names(tube_test_80)[5] <- "Sex"
names(tube_test_80)[6] <- "Exp_group"

#Transform the dataframe from a wide to a long format
tube_test_80 <- melt(tube_test_80, id.vars = c("Name","Arrive_weight", "Phenotype", "Crate", "Sex", "Exp_group","Cage.ID"))

#Renaming the columns created by melt()
names(tube_test_80)[8] <- "Treatments"
names(tube_test_80)[9] <- "Wins_per_rep"
names(tube_test_80)[7] <- "Cage_ID"
names(tube_test_80)[4] <- "Crate_ID"

#Add a new column with the replicate info and then rename the rows of column Treatments to only the treatments 
Replicates <- c(rep("R1",80), rep("R2",80), rep("R3",80),rep("R4",80),rep(c(rep("R1",80), rep("R2",80), rep("R3",80)),4),rep("R3",80), rep("R4",80), rep("R5",80))
#add the column "Replicates" to the dataframe tube_test_80
tube_test_80 <- cbind(tube_test_80,Replicates)
#Changing the name of the factors created after transforming the dataframe. There are 19 in total corresponding to the no of reps in total.
levels(tube_test_80$Treatments) <- c("OPT","OPT","OPT","OPT","HEAT","HEAT","HEAT","COLD","COLD","COLD","DIET","DIET","DIET","ANTI","ANTI","ANTI","FMT","FMT","FMT")
levels(tube_test_80$Exp_group) <- c("Treatment", "Control") #does the glm or lm model then know how to differ from the experimental and control animals tube test wins? Ask about this.

#Move the no of wins column (all of it) to the first column
colnames(tube_test_80)
tube_test_80 <- tube_test_80[, c(9,8,10,1:7)]
```

## Dataframe transformation

The files have been read in and are now transformed. I create two dataframes, where tube_test_80 include all 80 mice and tube_test_40 include 40 mice. These 40 mice are the ones that we have genomics and metagenomics data on. The 40 mice are composed of 20 females and 20 males: The male cages are 3, 4, 5 and 6. The female cages are 10, 11, 12 and 13. Cage 4 (male) and 13 (female) are control cages.

```{r}
#Creating new factors to start statistics. 
#A factor instead of num, chr or int account for variation between the "categories"
tube_test_80 <- transform(tube_test_80,Replicates=as.factor(Replicates),
                            Phenotype=as.factor(Phenotype),
                            Crate=as.factor(Crate_ID),
                            Sex=as.factor(Sex),
                            Exp_group=as.factor(Exp_group),
                            Cage_ID=as.factor(Cage_ID))
str(tube_test_80) #check if the command worked
head(tube_test_80)

#Remove mice that are not among the 40 I'm working with:
#I have mice from cage 3-6 (males) and 10-13 (females)
cages_to_remove <- c("01M","02M","07M","08M","09F","14F","15F","16F")
condition <- tube_test_80$Cage_ID %in% cages_to_remove
tube_test_40 <- tube_test_80[!condition, ]
# Remove rows with missing values (NA) in any column if needing the df like this later
cleaned_tube_test_40 <- drop_na(tube_test_40)

str(tube_test_40)
head(tube_test_40)
```

## Histograms of the dataframes

I plot histograms of both dataframes to visualize the distribution of total number of tube test wins. The data (number of wins) is not normal distributed, but this was expected.

```{r, echo=FALSE, fig.align='center',out.width='70%',fig.height=10}
par(mfrow=c(2,1))  # Arrange the plots in two rows

#hist(tube_test_80$Wins_per_rep) #why gaps?
#hist(tube_test_40$Wins_per_rep) #why gaps?
#I can't figure out why the plots looks weird with spaces/gaps in between the bars?

#So far this is the solution:
hist(tube_test_80$Wins_per_rep, breaks=seq(0,4,by=0.2), ylim = c(0,400), xlab="Tube test wins", main="Histogram showing frequency of wins for the 80 mice")
hist(tube_test_40$Wins_per_rep, breaks=seq(0,4,by=0.2), ylim = c(0,200),xlab="Tube test wins", main="Histogram showing frequency of wins for the 40 mice")
```

## Statistics using lm() and glm() 

I have done some statistical tests using linear models and general linear models to test the independent variables against the dependable variable (dominane, number of wins). Many of the tests might not make sense... But perhaps you Inaki have some advice on which ones to use? Because I think glm() works, I just have problems interpreting the results. I hope you can help with this.

First I glimpse at the data:

```{r}
glimpse(tube_test_40) 
glimpse(tube_test_80) 
```

Here I test the data using lm()

```{r}
#Because the Mice IDs are not of relevance here, we exclude the Name column, 
#but test all other variables/coefficients/predictors for both df: tube_test_80 and tube_test_40
lm_result_all <- lm(Wins_per_rep ~ .-Name,data=tube_test_80)
summary(lm_result_all)
plot(lm_result_all)

lm_result_all <- lm(Wins_per_rep ~ .-Name,data=tube_test_40)
summary(lm_result_all)
plot(lm_result_all)
```

Here I test the data using glm()

```{r}
#Because the Mice IDs are not of relevance here, we exclude the Name column, 
#but test all other variables/coefficients/predictors for both df: tube_test_80 and tube_test_40
glm_result <- glm(Wins_per_rep ~ .-Name,data=tube_test_80,family = poisson())
summary(glm_result)
plot(glm_result)

glm_result <- glm(Wins_per_rep ~ .-Name,data=tube_test_40,family = poisson())
summary(glm_result)
plot(glm_result)
```

I read about testing families and testing log transformed data using glm(). I tried to test different things, but not all codes work for me. If you Inkaki find this relevant, then I would like some help. 

```{r}
#However, other families may be a better fit for my data, and thus, I read about them using ?glm()
#I also used ChatGPT to help me understand the definition of each family 
#Turns out the Quasipoisson could fit my data better than poisson
#So to find out if this is the case, I have to explore my data
#If the  variance is larger than the mean in my dataset (=overdispersion), then quasipoisson is a better fit
#You only calculate the variance (sd) and mean of the dependent variable, which in my case is number of wins

#First testing this with all 80 mice
dependent_variable_wins <- tube_test_80$Wins_per_rep
mean_value <- mean(dependent_variable_wins,na.rm=TRUE)
sd_value <- sd(dependent_variable_wins,na.rm=TRUE)

coefficient_dispersion <- (sd_value / mean_value) * 100
coefficient_variation <- (sd_value / mean_value) * 100

print(coefficient_dispersion) 
print(coefficient_variation)
#The results of the 40 specific mice (tube_test_40) were: 68.8126 and 68.8126
#In this case, the standard deviation (sd) = variance is equal to the mean, resulting in the coefficients being the same.
#So since my data is count data with no overdispersion, the Poisson family might be suitable indeed.
#When the standard deviation is equal to the mean, it implies that the data values are evenly distributed around the mean (which is clear when visualising the number of wins in a histogram). 
#This can occur when the data is symmetrically distributed or when there is a specific relationship between the values in your dataset. The latter is my data (0,1,2,3 or 4 wins)

#I looked at papers doing tube test analyses and one named "social hierarchy position in female mice is associated with plasma corticosterone levels and hypothalamic gene expression" (2019) by Williamson et al. uses a Poisson-lognormal distribution test.
#I want to see if the Poisson-lognormal distribution test fits my data better than regular poisson

#TEST ON SPECIFIC 40 MICE
# Fit Poisson regression model
#poisson_model <- glm(Wins_per_rep ~ . - Name, data = cleaned_tube_test_40, family = "poisson")
poisson_model <- glm(Wins_per_rep ~ Treatments + Cage_ID, data = cleaned_tube_test_40, family = "poisson")

# Fit Poisson-lognormal regression model with log-transformed response
#log_count_var <- log(cleaned_tube_test_40$Wins_per_rep)
#cleaned_tube_test_40$log_wins <- log_count_var
#head(cleaned_tube_test_40)
#poisson_lognormal_model <- glm(log_wins ~ . - Name, data = cleaned_tube_test_40, family = "gaussian") #does not work
#poisson_lognormal_model <- glm(log_wins ~ Treatments + Cage_ID, data = cleaned_tube_test_40, family = "gaussian") #does not work
#cleaned_tube_test_40 <- within(cleaned_tube_test_40, rm(log_wins))
#poisson_lognormal_model <- glm(log(Wins_per_rep) ~ . - Name, data = cleaned_tube_test_40, family = "gaussian") #does not work
#poisson_lognormal_model <- glm(log(Wins_per_rep) ~ Treatments + Cage_ID, data = cleaned_tube_test_40, family = "gaussian") #does not work
```

## Visualizations based on simple questions

Antton and Ostaizka recommened me to ask simple questions and try to answer those via visualizations. They are in this section asked and then the associated plots are my attempts of answering them. I showed you some of them them in the presentation end August. I would love to get your inputs Inaki; if you think I answer the questions by choosing the plots I did, or if you think I should visualize the data differently.

FIRST QUESTION:
1. What is the dominance hierarchy per cage for each treatments? Is the hierarchy stable/the same across tube test replicates?

I try to answer this by visualizing the absolute number of wins per mouse per replicate for the OPT and FMT treatments. I have done similar plots of the other treatment, but do not show them here. If you Inaki want me to, then please let me know. 

```{r,echo=FALSE,fig.align='center',out.width='70%',fig.height=5}
par(mar=c(6,4,4,11), xpd=TRUE)

opt_wins <- tube_test_wins[,c(1,5, 11:17)]
## Loop creating plots for the absolute wins per cage in the 4 replicates of OPT
i<-1
while (i!=81){
  matplot(t(opt_wins[c(i:(i+4)),c(3:6)]),type="b",xaxt = "n",xlab = "Replicate", ylab= "n of wins", ylim = c(0,4.5), pch=20)
  axis(1,at =c(1,2,3,4),labels=c("R1","R2","R3","R4"))
  nn <- 5
  legend(x= 4.3, y=3, opt_wins$Name[i:(i+4)], col=seq_len(nn), cex=0.6,fill=seq_len(nn),  bty= "n", xpd=TRUE)    
  title(paste("Absolute wins of ", opt_wins$Cage[i], " in OPT per replicate"))
  i<-i+5
}

fmt_wins <- tube_test_wins[,c(1,5, 36:38,40:41)]
## Loop creating plots for the absolute wins per cage in the 5 replicates of FMT ##
i<-1
while (i!=81){
  matplot(t(fmt_wins[c(i:(i+4)),c(3:7)]),type="b",xaxt = "n",xlab = "Replicate", ylab= "n of wins", ylim = c(0,4.5), pch=20)
  axis(1,at =c(1,2,3,4,5),labels=c("R1","R2","R3","R4","R5"))
  nn <- 5
  legend(x= 5.3, y=3, fmt_wins$Name[i:(i+4)], col=seq_len(nn), cex=0.6, fill=seq_len(nn), bty= "n", xpd=TRUE)    
  title(paste("Absolute wins of ", fmt_wins$Cage[i], " in FMT per replicate"))
  i<-i+5
}
```

ANSWER TO THE FIRST QUESTION:
The dominance hierarchy (the replicates) changes within a treatment per cage. Hence, no the respective cage dominance hierarchies are not stable. They change within tube test replicates for each treatment.



SECOND QUESTION:
2. Is there an overall or stable hierarchy (despite treatments)? For example one mouse continuously winning in the experiment.

I try to answer this by visualizing the average number of wins per mouse per treatment as dot plots.

```{r, echo=FALSE,fig.align='center',out.width='70%',fig.height=5}
avg_wins <- tube_test_wins[,c(1,5, 15, 22, 46:49)]
avg_wins$Cage <- factor(avg_wins$Cage, levels=unique(avg_wins$Cage)) #to keep the order of the cages
avg_wins_by_cage= split(avg_wins, avg_wins$Cage)

par(mar=c(5,4,4,9), xpd=TRUE)

for (i in avg_wins_by_cage){
  matplot(t(i[,c(3:8)]),type="b", ,xaxt = "n", ylab="Mean wins",pch=20, ylim=c(0,4))
  axis(1,at =c(1,2,3,4,5,6),labels=c("OPT", "HEAT", "COLD", "DIET", "ANTI", "FMT"))
  nn <- 5
  legend(x= 6.4, y=3, cex=0.55, i$Name, col=seq_len(nn), fill=seq_len(nn), bty= "n", xpd=TRUE)    
  title(paste("Average wins of each mouse of ", i$Cage[1], " per treatment"))
}
```

I also try to answer this by visualizing the average number of wins per mouse per treatment as violin plots. If the violin plot look identical, then the dominance hierachy is stable.

```{r, echo=FALSE,fig.align='center',out.width='70%',fig.height=5}
total_avg_wins <- tube_test_wins[,c(1,5,6,50)]
total_avg_wins_40 <- total_avg_wins[-c(1:10,31:45,66:80),] #removing irrelevant rows/cages
            ### Note to self: change column names of total_avg_wins_40 e.g. Average.rank..wins..for.all.treatments
#total_avg_wins$Cage <- factor(total_avg_wins$Cage, levels=unique(total_avg_wins$Cage)) #to keep the order of the cages
#total_avg_wins_by_cage= split(total_avg_wins, total_avg_wins$Cage)

#Does this take individuality/mouse ID into account? YES! It does. Look at notes on my Mac.
avg_wins <- tube_test_wins[,c(1,5, 15, 22, 46:49)]
avg_wins_40 <- avg_wins[-c(1:10,31:45,66:80),] #removing irrelevant cages
OPT <- avg_wins_40$Rank.of.Wins.OPT.Avg
HEAT <- avg_wins_40$Rank.of.Wins.HEAT.Avg
COLD <- avg_wins_40$Rank.of.wins.COLD.Avg
DIET <- avg_wins_40$Rank.of.wins.DIET.Avg
ANTI <- avg_wins_40$Rank.of.wins.ANTI.Avg
FMT <- avg_wins_40$Rank.of.wins.FMTs.3.5.Avg
vioplot(OPT, HEAT, COLD, DIET, ANTI, FMT, names=c("OPT", "HEAT", "COLD", "DIET", "ANTI", "FMT"),  col=2:7)
title("Violin Plots showing the distribution of 40 mice's mean wins per treatment")
#if the domiance rank/hierarchy is stable, then the median of the means would be the same, and the violins would look the same too)

# Create the basic ggplot object
plot <- ggplot(total_avg_wins, aes(x = Name, y = Average.rank..wins..for.all.treatments, fill = Cage)) +
  geom_bar(stat = "identity") +
  labs(x = "Name", y = "Average Wins", title = "Average Wins for All Treatments") +
  scale_fill_discrete(guide = "none") +  # Remove legend
  theme_minimal() +  # Set a transparent background
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

Moreover, I tried to answer the question by looking at correlations. I use the mean number of wins per mouse, and it should be individual specific. But please have a look at the code Inaki and let me know.

```{r, echo=FALSE,fig.align='center',out.width='70%',fig.height=8}

plotting_all <- tube_test_wins[,c(1,9,15,22,46:49)]
plotting_all_matrix <- data.matrix(plotting_all) #turn the dataframe into a matrix

chart.Correlation(plotting_all_matrix[,c(3:8)], histogram=TRUE, pch=19)

#ggscatter(plotting_all, x = "Rank.of.Wins.OPT.Avg", y = "Rank.of.wins.FMTs.3.5.Avg", 
#               add = "reg.line", conf.int = TRUE, 
#               cor.coef = TRUE, cor.method = "pearson",
#               xlab = "OPT wins for all mice", ylab = "HEAT wins for all mice", color = "Rank.of.wins.FMTs.3.5.Avg")
```

ANSWER TO THE SECOND QUESTION:
There is not a stable hierarchy for each cage across treatments. The hierarchy changes for all cages. However, e.g. cage 13 has one mouse dominating from the beginning to the end - on average winning 4 times per treatment. 

Nevertheless, it appears that there's an overall hierarchy for each cage across treatments. Thus, that time matters. The correlations on mean wins per mouse show that the treatments close to each other correlate. I was recommended by Antton and Ostaizka to split the data into e.g. sex and cages, and then do tests, correlation plots etc. Do you agree? Either way, I think it would be smart to do. So I'll work on it this week.



THIRD QUESTION:
3. Is there a difference in dominance between the sex?

ANSWER TO THE THIRD QUESTION:
I feel like I've answered this question in the statistical tests using glm() that shows "NA" at sex, which means no significance. But perhaps you Inaki can suggest something?



FOURTH QUESTION:
4. What does the accumulative number of wins look like per mouse within each cage? 

```{r,echo=FALSE,fig.align='center',out.width='70%',fig.height=10}
plot <- ggplot(total_avg_wins, aes(x = Name, y = Average.rank..wins..for.all.treatments, fill = Cage)) + geom_bar(stat = "identity") +
  labs(x = "Name", y = "Average Wins", title = "Average Wins for All Treatments") +
  scale_fill_discrete(guide = "none") +  # Remove legend
  theme_minimal() +  # Set a transparent background
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
```

ANSWER TO THE FOURTH QUESTION:

We can see that one mouse per cage appears to dominate across the experiment despite treatments according to accumulative number of wins.



I look forward to hear what you think of it all Inaki. Thanks a lot for helping me.